File: packages/emitter/testcases/functions/type-guards/TypeGuards.ts
export interface Dog {
  type: "dog";
  bark(): void;
}

export interface Cat {
  type: "cat";
  meow(): void;
}

export type Animal = Dog | Cat;

// Type guard function with 'is' predicate
export function isDog(animal: Animal): animal is Dog {
  return animal.type === "dog";
}

export function isCat(animal: Animal): animal is Cat {
  return animal.type === "cat";
}

// Using type guards
export function makeSound(animal: Animal): void {
  if (isDog(animal)) {
    animal.bark();
  } else if (isCat(animal)) {
    animal.meow();
  }
}

// typeof type guard
export function processValue(value: string | number): string {
  if (typeof value === "string") {
    return value.toUpperCase();
  }
  return value.toString();
}

// instanceof type guard
export class Circle {
  radius: number;
  constructor(radius: number) {
    this.radius = radius;
  }
}

export function getArea(shape: Circle | number): number {
  if (shape instanceof Circle) {
    return Math.PI * shape.radius * shape.radius;
  }
  return shape;
}

File: packages/emitter/testcases/functions/type-guards/TypeGuards.cs
namespace TestCases.functions.typeguards
{
    public class Dog
    {
        public string type { get; set; }

        public void bark() => throw new NotImplementedException();
    }
    public class Cat
    {
        public string type { get; set; }

        public void meow() => throw new NotImplementedException();
    }
    public class Circle
    {
        public double radius;

        public Circle(double radius)
            {
            this.radius = radius;
            }
    }

            public static class TypeGuards
            {
                // type Animal = global::Tsonic.Runtime.Union<Dog, Cat>

                public static bool isDog(Animal animal)
                    {
                    return animal.type == "dog";
                    }

                public static bool isCat(Animal animal)
                    {
                    return animal.type == "cat";
                    }

                public static void makeSound(Animal animal)
                    {
                    if (isDog(animal))
                        {
                        animal.bark();
                        }
                    else
                        if (isCat(animal))
                            {
                            animal.meow();
                            }
                    }

                public static string processValue(global::Tsonic.Runtime.Union<string, double> value)
                    {
                    if (global::Tsonic.Runtime.Operators.@typeof(value) == "string")
                        {
                        return global::Tsonic.JSRuntime.String.toUpperCase(value);
                        }
                    return global::Tsonic.JSRuntime.Number.toString(value);
                    }

                public static double getArea(global::Tsonic.Runtime.Union<Circle, double> shape)
                    {
                    if (shape is Circle)
                        {
                        return global::Tsonic.JSRuntime.Math.PI * shape.radius * shape.radius;
                        }
                    return shape;
                    }
            }
}